### PR TITLE
Add "expand" command to Crystal tools

### PR DESCRIPTION
**Pull request by MakeNowJust**
`crystal tool expand` expands macro for given location.

See asciinemas:

[asciicast](https://asciinema.org/a/96590)

[asciicast](https://asciinema.org/a/96600)

### CODE REVIEW

**Comment by bcardiff** [comment]
Nice. I would like to see some specs. At least some basic one that will ensure some scenarios wont fail: nested classes with modules, top level expansions.

I bet you wish there was some kind of index per file/line so you don't need to lookup the whole hierarchy of types. For now we are good, but in the future we might need to enhance de compiler with some of those indices i think.

**Comment by bcardiff** [comment]
Regarding the specs one that will result in multiple expansions should be interesting.

And json support with an array of the different levels of expansions. So we can zoom in/out. WDYT?

**Comment by MakeNowJust** [comment]
@bcardiff 

> And json support with an array of the different levels of expansions. So we can zoom in/out. WDYT?

When macros are nested, final expansion result may be too complex. In this case it is useful.

> Regarding the specs one that will result in multiple expansions should be interesting.

If there is a macro in a method definition and the method definition has multiple type signatures, the macro has multiple expansion. For example:

```crystal
macro foo
  puts "hello"
end

def bar(x)
  foo
end

bar
```

`./bin/crystal tool expand -c foo.cr:6:3 foo.cr` yields:

```
﻿2 expansions found
expansion 1:
   foo

~> puts("hello")

expansion 2:
   foo

~> puts("hello")

```

- - -

I'll try to write specs. But, it is midnight in Japan, I am going to bed. Good night!

**Comment by MakeNowJust** [comment]
tiny `.vimrc` for demos:

```viml
command! -buffer -nargs=0 CrystalExpand echo s:crystal_expand(expand('%'), getpos('.'))

function! s:crystal_expand(file, pos)
  let l:cmd = printf('./bin/crystal tool expand --no-color -c %s:%d:%d %s', a:file, a:pos[1], a:pos[2], a:file)
  return system(l:cmd)
endfunction
```

**Comment by RX14** [comment]
Could we run the output through crystal format? Some of the generated macro code isn't particularly beautiful.

**Comment by asterite** [comment]
@MakeNowJust Thank you for this! I think this tool can be useful, I just need to think what's the criteria to merge new tools.

Regardless of that, I find it really cool that you are hacking the compiler :-)

File: src/compiler/crystal/tools/context.cr [hunk 1]
```diff
@@ -289,9 +292,10 @@ module Crystal
     private def contains_target(node)
       if loc_start = node.location
         loc_end = node.end_location || loc_start
-        # if it is not between, it could be the case that node is the top level Expressions
+        # if it is not between, it could be the case that node is the top level Expressions,
         # in which the (start) location might be in one file and the end location in another.
-        @target_location.between?(loc_start, loc_end) || loc_start.filename != loc_end.filename
+        @target_location.between?(loc_start, loc_end) || (!@in_defs && loc_start.filename != loc_end.filename)
+        # When node generated by macro and macro has yielding block, it also makes `loc_start.filename != loc_end.filename` true.
```
Comment (bcardiff) [inline src/compiler/crystal/tools/context.cr]: Great! Can a spec be added also for this case? The specs are at `./spec/compiler/crystal/tools/context_spec.cr` and should be easy to add a sample. There are some unicode chars that act as user cursor so you can easily add specs.

Comment (MakeNowJust) [inline src/compiler/crystal/tools/context.cr]: It is just ad-hoc solution. The radical problem is `yield` in macro expansion sets its location, but doesn't reset it. Now, I have an idea for it. I'll make a new pull request to solve it and I am going to revert this commit after this pull request is merged.

Comment (MakeNowJust) [inline src/compiler/crystal/tools/context.cr]: It is fixed in #3751. I'll revert and rebase.

**Comment by MakeNowJust** [comment]
@bcardiff Added specs just now!

File: src/compiler/crystal/tools/expand.cr [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+          io.puts
+        end
+      end
+    end
+  end
+
+  class ExpandTrace
+    JSON.mapping({
+      original:  {type: String},
+      expandeds: {type: Array(String)},
```
Comment (MakeNowJust) [inline src/compiler/crystal/tools/expand.cr]: Yes. I think `expansions` is better.

File: src/compiler/crystal/tools/expand.cr [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    })
+
+    def initialize(@original, @expandeds)
+    end
+
+    def self.ast_to_s(node)
+      s = String.build { |io| node.to_s(io, emit_doc: true) }
+      return s unless node.is_a?(MacroIf) || node.is_a?(MacroFor)
+      indent = node.location.not_nil!.column_number - 1
```
Comment (MakeNowJust) [inline src/compiler/crystal/tools/expand.cr]: `MacroIf` and `MacroFor` have `MacroBody`, it is just sub-string of source text. For example:

```crystal
begin
  {% if 1 == 1 %}
    hello
  {% end %}
end
```

`MacroIf#to_s` result of `{% if 1 == 1 %}...{% end %}` is:

```crystal
{% if 1 == 1 %}
    hello
  {% end %}
```

So, re-indentation is needed. I'll add specs for it.

File: src/compiler/crystal/tools/expand.cr (hunk 2 above)
Comment (MakeNowJust) [inline src/compiler/crystal/tools/expand.cr]: Hmm, but we will get `expansions.expansions`. It is very wrong.

**Comment by MakeNowJust** [comment]
@bcardiff Updated.

File: src/compiler/crystal/tools/expand.cr (hunk 2 above)
Comment (mverzilli) [inline src/compiler/crystal/tools/expand.cr]: `expanded` is also used to address a plural quantity, so you can just `s/expandeds/expanded` :)

**Comment by bcardiff** [comment]
🎉 🔨 🔧 Thanks again @MakeNowJust  🔧 🔨 🎉

**Pull request merged**